Q1. Notebook
We'll start with the same notebook we ended up with in homework 1. We cleaned it a little bit and kept only the scoring part. You can find the initial notebook here.

Run this notebook for the March 2023 data.

What's the standard deviation of the predicted duration for this dataset?

1.24
6.24
12.28
18.28

In [11]:
#code to load in the Mar-2023 data
import pickle
import pandas as pd

with open('model.bin', 'rb') as f_in:
    dv, model = pickle.load(f_in)

categorical = ['PULocationID', 'DOLocationID']

def read_data(filename):
    df = pd.read_parquet(filename)
    
    df['duration'] = df.tpep_dropoff_datetime - df.tpep_pickup_datetime
    df['duration'] = df.duration.dt.total_seconds() / 60

    df = df[(df.duration >= 1) & (df.duration <= 60)].copy()

    df[categorical] = df[categorical].fillna(-1).astype('int').astype('str')
    
    return df

df = read_data('https://d37ci6vzurychx.cloudfront.net/trip-data/yellow_tripdata_2023-03.parquet')
df.head()

/home/codespace/.local/lib/python3.12/site-packages/sklearn/base.py:380: InconsistentVersionWarning: Trying to unpickle estimator DictVectorizer from version 1.5.0 when using version 1.6.1. This might lead to breaking code or invalid results. Use at your own risk. For more info please refer to:
https://scikit-learn.org/stable/model_persistence.html#security-maintainability-limitations
  warnings.warn(
/home/codespace/.local/lib/python3.12/site-packages/sklearn/base.py:380: InconsistentVersionWarning: Trying to unpickle estimator LinearRegression from version 1.5.0 when using version 1.6.1. This might lead to breaking code or invalid results. Use at your own risk. For more info please refer to:
https://scikit-learn.org/stable/model_persistence.html#security-maintainability-limitations
  warnings.warn(


VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0         2  2023-03-01 00:06:43   2023-03-01 00:16:43              1.0   
1         2  2023-03-01 00:08:25   2023-03-01 00:39:30              2.0   
2         1  2023-03-01 00:15:04   2023-03-01 00:29:26              0.0   
3         1  2023-03-01 00:49:37   2023-03-01 01:01:05              1.0   
4         2  2023-03-01 00:08:04   2023-03-01 00:11:06              1.0   

   trip_distance  RatecodeID store_and_fwd_flag PULocationID DOLocationID  \
0           0.00         1.0                  N          238           42   
1          12.40         1.0                  N          138          231   
2           3.30         1.0                  N          140          186   
3           2.90         1.0                  N          140           43   
4           1.23         1.0                  N           79          137   

   payment_type  fare_amount  extra  mta_tax  tip_amount  tolls_amount  \
0             2          8.6    1.0      0.5        0.00           0.0   
1             1         52.7    6.0      0.5       12.54           0.0   
2             1         18.4    3.5      0.5        4.65           0.0   
3             1         15.6    3.5      0.5        4.10           0.0   
4             1          7.2    1.0      0.5        2.44           0.0   

   improvement_surcharge  total_amount  congestion_surcharge  Airport_fee  \
0                    1.0         11.10                   0.0         0.00   
1                    1.0         76.49                   2.5         1.25   
2                    1.0         28.05                   2.5         0.00   
3                    1.0         24.70                   2.5         0.00   
4                    1.0         14.64                   2.5         0.00   

    duration  
0  10.000000  
1  31.083333  
2  14.366667  
3  11.466667  
4   3.033333

In [12]:
dicts = df[categorical].to_dict(orient='records')
X_val = dv.transform(dicts)
y_pred = model.predict(X_val)

In [13]:
#Now get the standard deviation for the duration
import numpy as np
print(f"The standard deviation of the predicted durations is {np.std(y_pred):.2f}")

The standard deviation of the predicted durations is 6.25


The standard deviation of the predicted durations is 6.25
So the closest answer is 6.24

Q2. Preparing the output
Like in the course videos, we want to prepare the dataframe with the output.

First, let's create an artificial ride_id column:

df['ride_id'] = f'{year:04d}/{month:02d}_' + df.index.astype('str')
Next, write the ride id and the predictions to a dataframe with results.

Save it as parquet:

df_result.to_parquet(
    output_file,
    engine='pyarrow',
    compression=None,
    index=False
)
What's the size of the output file?

36M
46M
56M
66M
Note: Make sure you use the snippet above for saving the file. It should contain only these two columns. For this question, don't change the dtypes of the columns and use pyarrow, not fastparquet.

In [14]:
year = 2023
month = 3
df['ride_id'] = f'{year:04d}/{month:02d}_' + df.index.astype('str')
df.head()

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0         2  2023-03-01 00:06:43   2023-03-01 00:16:43              1.0   
1         2  2023-03-01 00:08:25   2023-03-01 00:39:30              2.0   
2         1  2023-03-01 00:15:04   2023-03-01 00:29:26              0.0   
3         1  2023-03-01 00:49:37   2023-03-01 01:01:05              1.0   
4         2  2023-03-01 00:08:04   2023-03-01 00:11:06              1.0   

   trip_distance  RatecodeID store_and_fwd_flag PULocationID DOLocationID  \
0           0.00         1.0                  N          238           42   
1          12.40         1.0                  N          138          231   
2           3.30         1.0                  N          140          186   
3           2.90         1.0                  N          140           43   
4           1.23         1.0                  N           79          137   

   payment_type  ...  extra  mta_tax  tip_amount  tolls_amount  \
0             2  ...    1.0      0.5        0.00           0.0   
1             1  ...    6.0      0.5       12.54           0.0   
2             1  ...    3.5      0.5        4.65           0.0   
3             1  ...    3.5      0.5        4.10           0.0   
4             1  ...    1.0      0.5        2.44           0.0   

   improvement_surcharge  total_amount  congestion_surcharge  Airport_fee  \
0                    1.0         11.10                   0.0         0.00   
1                    1.0         76.49                   2.5         1.25   
2                    1.0         28.05                   2.5         0.00   
3                    1.0         24.70                   2.5         0.00   
4                    1.0         14.64                   2.5         0.00   

    duration    ride_id  
0  10.000000  2023/03_0  
1  31.083333  2023/03_1  
2  14.366667  2023/03_2  
3  11.466667  2023/03_3  
4   3.033333  2023/03_4  

[5 rows x 21 columns]

In [15]:
df_result = pd.DataFrame({
    'ride_id': df['ride_id'],
    'predicted_duration': y_pred
})
df_result.tail()

ride_id  predicted_duration
3403761  2023/03_3403761           11.952635
3403762  2023/03_3403762           20.049958
3403763  2023/03_3403763           11.595336
3403764  2023/03_3403764           13.113178
3403765  2023/03_3403765           12.899992

In [16]:
#save as parquet
df_result.to_parquet(
    'df_result.parquet',
    engine='pyarrow',
    compression=None,
    index=False
)

In [ ]:
!ls -lh df_result.parquet

-rw-rw-rw- 1 codespace codespace 66M Jun 14 07:43 df_result.parquet


: 

 answer is 66MB

Q3. Creating the scoring script
Now let's turn the notebook into a script.

Which command you need to execute for that?

Answer
In the command line

jupyter nbconvert --to python starter.ipynb 

Q4. Virtual environment
Now let's put everything into a virtual environment. We'll use pipenv for that.

Install all the required libraries. Pay attention to the Scikit-Learn version: it should be the same as in the starter notebook.

After installing the libraries, pipenv creates two files: Pipfile and Pipfile.lock. The Pipfile.lock file keeps the hashes of the dependencies we use for the virtual env.

What's the first hash for the Scikit-Learn dependency?

Here are my steps for creating the venv

Create the pipenv
pipenv install scikit-learn==1.5.0 flask --python=3.9
Active/set into the environment
pipenv shell
Open the pipfile.lock to find the first hash. In this case it is sha256:0e8102d5036e28d08ab47166b48c8d5e5810704daecf3a476a4282d562be9a28


Found I needed a few other libraries Added in

pipenv install pandas pyarrow
pipenv lock

Q5. Parametrize the script
Let's now make the script configurable via CLI. We'll create two parameters: year and month.

Run the script for April 2023.

What's the mean predicted duration?

7.29
14.29
21.29
28.29
Hint: just add a print statement to your script.

Starter.py script is now

import pickle
import pandas as pd
import os

with open('model.bin', 'rb') as f_in:
    dv, model = pickle.load(f_in)

categorical = ['PULocationID', 'DOLocationID']

YEAR = os.getenv('YEAR')
MONTH = os.getenv('MONTH')

def read_data(filename):
    df = pd.read_parquet(filename)
    
    df['duration'] = df.tpep_dropoff_datetime - df.tpep_pickup_datetime
    df['duration'] = df.duration.dt.total_seconds() / 60

    df = df[(df.duration >= 1) & (df.duration <= 60)].copy()

    df[categorical] = df[categorical].fillna(-1).astype('int').astype('str')
    
    return df

df = read_data('https://d37ci6vzurychx.cloudfront.net/trip-data/yellow_tripdata_{YEAR}-{MONTH}.parquet')

dicts = df[categorical].to_dict(orient='records')
X_val = dv.transform(dicts)
y_pred = model.predict(X_val)

df['ride_id'] = f'{YEAR:04d}/{MONTH:02d}_' + df.index.astype('str')
df_result = pd.DataFrame({
    'ride_id': df['ride_id'],
    'predicted_duration': y_pred
})
df_result.to_parquet(
    'df_result.parquet',
    engine='pyarrow',
    compression=None,
    index=False
)

#print the mean predicted duration
print(f"The mean predicted duration is {df_result['predicted_duration'].mean():.2f}")
So in the CLI I typed

export YEAR=2023
export MONTH=04
python starter.py

Q6. Docker container
Finally, we'll package the script in the docker container. For that, you'll need to use a base image that we prepared.

This is what the content of this image is:

FROM python:3.10.13-slim

WORKDIR /app
COPY [ "model2.bin", "model.bin" ]
Note: you don't need to run it. We have already done it.

It is pushed it to agrigorev/zoomcamp-model:mlops-2024-3.10.13-slim, which you need to use as your base image.

That is, your Dockerfile should start with:

FROM agrigorev/zoomcamp-model:mlops-2024-3.10.13-slim

This image already has a pickle file with a dictionary vectorizer and a model. You will need to use them.

Important: don't copy the model to the docker image. You will need to use the pickle file already in the image.

Now run the script with docker. What's the mean predicted duration for May 2023?

0.19
7.24
14.24
21.19
The answer is 0.19